pkg/didcomm/crypto/legacy/authcrypt/authcrypt.go

```diff
@@ -0,0 +1,224 @@
+// Crypter represents an Authcrypt Encrypter (Decrypter) that outputs/reads legacy Aries envelopes
+type Crypter struct {
+	sender     keyPairEd25519
+	recipients []*publicEd25519
+	alg        contentEncryption
+	nonceSize  int
+}
+
+// New will create a Crypter that encrypts messages using the legacy Aries format
+// Note: legacy crypter does not support XChacha20Poly1035 (XC20P), only Chacha20Poly1035 (C20P)
+func New(sender keyPairEd25519, recipients []*publicEd25519, alg contentEncryption) (*Crypter, error) {
+	var nonceSize int
+	switch alg {
+	case C20P:
+		nonceSize = chacha.NonceSize
+	default:
+		return nil, fmt.Errorf("encryption algorithm '%s' not supported", alg)
+	}
+
+	if len(recipients) == 0 {
+		return nil, errors.New("empty recipients keys, must have at least one recipient")
+	}
+	var recipientsKey []*publicEd25519
+	recipientsKey = append(recipientsKey, recipients...)
+
+	c := &Crypter{
+		sender:     sender,
+		recipients: recipientsKey,
+		alg:        alg,
+		nonceSize:  nonceSize,
+	}
+
+	if !isKeyPairValid(sender) {
+		return nil, fmt.Errorf(
+			"sender keyPair not supported, it must have a %d byte private key and %d byte public key",
+			ed25519.PrivateKeySize,
+			ed25519.PublicKeySize)
+	}
+
+	return c, nil
+}
+
+func isKeyPairValid(kp keyPairEd25519) bool {
+	if kp.priv == nil || kp.pub == nil {
+		return false
+	}
+
+	return true
+}
+
+// Envelope is the full payload envelope for the JSON message
+type Envelope struct {
+	Protected  string `json:"protected,omitempty"`
+	IV         string `json:"iv,omitempty"`
+	CipherText string `json:"ciphertext,omitempty"`
+	Tag        string `json:"tag,omitempty"`
+}
+
+// Protected is the protected header of the JSON envelope
+type Protected struct {
+	Enc        string      `json:"enc,omitempty"`
+	Typ        string      `json:"typ,omitempty"`
+	Alg        string      `json:"alg,omitempty"`
+	Recipients []Recipient `json:"recipients,omitempty"`
+}
+
+// Recipient holds the data for a recipient in the envelope header
+type Recipient struct {
+	EncryptedKey string          `json:"encrypted_key,omitempty"`
+	Header       RecipientHeader `json:"header,omitempty"`
+}
+
+// RecipientHeader holds the header data for a recipient
+type RecipientHeader struct {
+	KID    string `json:"kid,omitempty"`
+	Sender string `json:"sender,omitempty"`
+	IV     string `json:"iv,omitempty"`
+}
+
+// publicEd25519toCurve25519 takes an Ed25519 public key and provides the corresponding Curve25519 public key
+//  This function wraps PublicKeyToCurve25519 from Adam Langley's ed25519 repo: github.com/agl/ed25519
+func publicEd25519toCurve25519(pub *publicEd25519) (*publicCurve25519, error) {
+	pkOut := new([CurveKeySize]byte)
+	success := extra25519.PublicKeyToCurve25519(pkOut, (*[CurveKeySize]byte)(pub))
+	if !success {
+		return nil, errors.New("failed to convert public key")
+	}
+	return (*publicCurve25519)(pkOut), nil
+}
+
+// TODO: consider wrapping agl's secret key conversion instead of reimplementing
```

Comment:
is this TODO current? Can it be removed?

Reply:
That's something I want reviewed - I reimplemented a function with reference to agl's secret key conversion and libsodium's secret key conversion, since I didn't want the dependency. But I bit the bullet and imported agl's curve25519 package rather than implementing the _public_ key conversion, so maybe I should just turn this function into a wrapper or call the wrapped function directly as well.

Reply:
yes, we should call/wrap agl's library, not reimplement it.

Reply:
Done